File: sdks/python/nexus_sdk/transport/rpc.py
"""Native binary RPC transport — a thin wrapper around the Thunder client.

Thunder (``hivellm-thunder``) owns the wire (frame + MessagePack codec),
the single-connection multiplexer, and the handshake / reconnect /
credential re-send. This module adapts the SDK's value model
(``NexusValue``, capitalized kinds) to Thunder's (``Value``, lowercase
kinds) and preserves the ``Transport`` interface the rest of the SDK uses.

Handshake: the Nexus server ignores ``HELLO`` arguments and gates the
connection on a separate ``AUTH``, so Thunder's ``AUTH_COMMAND`` handshake
with the ``ARG_LESS`` HELLO style (bare ``HELLO []`` then ``AUTH [...]``)
matches the wire exactly and re-authenticates on reconnect.
"""

from __future__ import annotations

import asyncio
from typing import Optional

from thunder_rpc import AsyncClient, Config, Value
from thunder_rpc.client_config import ClientConfig, Credentials
from thunder_rpc.config import Handshake, HelloStyle

from nexus_sdk.transport.endpoint import Endpoint
from nexus_sdk.transport.types import (
    NexusValue,
    Transport,
    TransportCredentials,
    TransportRequest,
    TransportResponse,
    nx,
)


def _nexus_to_thunder(v: NexusValue) -> Value:
    """Translate the SDK's ``NexusValue`` into a Thunder ``Value``."""
    kind = v.kind
    if kind == "Null":
        return Value.null()
    if kind == "Bool":
        return Value.bool(v.value)
    if kind == "Int":
        return Value.int(v.value)
    if kind == "Float":
        return Value.float(v.value)
    if kind == "Bytes":
        return Value.bytes(v.value)
    if kind == "Str":
        return Value.str(v.value)
    if kind == "Array":
        return Value.array([_nexus_to_thunder(x) for x in v.value])
    if kind == "Map":
        return Value.map(
            [(_nexus_to_thunder(a), _nexus_to_thunder(b)) for a, b in v.value]
        )
    raise ValueError(f"unknown NexusValue kind: {kind!r}")


def _thunder_to_nexus(v: Value) -> NexusValue:
    """Translate a Thunder ``Value`` back into the SDK's ``NexusValue``."""
    kind = v.kind
    if kind == "null":
        return nx.Null()
    if kind == "bool":
        return nx.Bool(v.value)
    if kind == "int":
        return nx.Int(v.value)
    if kind == "float":
        return nx.Float(v.value)
    if kind == "bytes":
        return nx.Bytes(v.value)
    if kind == "str":
        return nx.Str(v.value)
    if kind == "array":
        return nx.Array([_thunder_to_nexus(x) for x in v.value])
    if kind == "map":
        return nx.Map(
            [(_thunder_to_nexus(a), _thunder_to_nexus(b)) for a, b in v.value]
        )
    raise ValueError(f"unknown Thunder Value kind: {kind!r}")


class RpcTransport(Transport):
    """Asyncio RPC transport backed by a single ``AsyncClient``."""

    def __init__(
        self,
        endpoint: Endpoint,
        credentials: TransportCredentials,
        connect_timeout_s: float = 5.0,
    ) -> None:
        self._endpoint = endpoint
        self._credentials = credentials
        self._connect_timeout_s = connect_timeout_s
        self._client: Optional[AsyncClient] = None
        self._connect_lock = asyncio.Lock()

    async def execute(self, req: TransportRequest) -> TransportResponse:
        client = await self._ensure_connected()
        # ``call`` returns the ``Ok`` payload directly and raises a typed
        # ``ThunderError`` on ``Err``; the SDK's higher layers surface those.
        result = await client.call(
            req.command, [_nexus_to_thunder(a) for a in req.args]
        )
        return TransportResponse(value=_thunder_to_nexus(result))

    def describe(self) -> str:
        return f"{self._endpoint} (RPC)"

    def is_rpc(self) -> bool:
        return True

    async def close(self) -> None:
        client = self._client
        self._client = None
        if client is not None:
            await client.close()

    # ── Internals ──────────────────────────────────────────────────────

    def _config(self) -> Config:
        return (
            Config.standard()
            .with_scheme("nexus")
            .with_port(self._endpoint.port)
            .with_handshake(Handshake.AUTH_COMMAND)
            .with_hello_style(HelloStyle.ARG_LESS)
        )

    def _thunder_credentials(self) -> Optional[Credentials]:
        if self._credentials.api_key:
            return Credentials.api_key(self._credentials.api_key)
        if self._credentials.username and self._credentials.password:
            return Credentials.user_pass(
                self._credentials.username, self._credentials.password
            )
        return None

    async def _ensure_connected(self) -> AsyncClient:
        if self._client is not None:
            return self._client
        async with self._connect_lock:
            if self._client is not None:
                return self._client
            # Bare ``host:port`` endpoint sidesteps Thunder's scheme matching.
            self._client = await AsyncClient.connect(
                f"{self._endpoint.host}:{self._endpoint.port}",
                self._config(),
                ClientConfig(
                    connect_timeout=self._connect_timeout_s,
                    credentials=self._thunder_credentials(),
                ),
            )
            return self._client

File: sdks/typescript/src/transports/rpc.ts
/**
 * Native binary RPC transport — a thin wrapper around the Thunder client.
 *
 * Thunder (`@hivehub/thunder`) owns the wire (frame + MessagePack codec,
 * SPEC-001), the single-connection multiplexer, and the handshake /
 * reconnect / credential re-send. This file adapts the SDK's own value
 * model (`NexusValue`, capitalized kinds) to Thunder's (`Value`, lowercase
 * kinds) and preserves the `Transport` interface the rest of the SDK
 * depends on.
 *
 * Handshake: the Nexus server ignores `HELLO` arguments and gates the
 * connection on a separate `AUTH`, so Thunder's `auth_command` handshake
 * with the `arg_less` HELLO style (bare `HELLO []` then `AUTH [...]`)
 * matches it exactly — and, unlike a hand-rolled connect, re-runs on
 * reconnect so credentials are re-sent automatically.
 */

import {
  Client,
  Config,
  Value,
  type Credentials,
} from '@hivehub/thunder';
import {
  Transport,
  TransportCredentials,
  TransportRequest,
  TransportResponse,
  NexusValue,
  nx,
} from './types';
import { Endpoint, endpointToString } from './endpoint';

/** Translate the SDK's `NexusValue` into a Thunder `Value`. */
function nexusToThunder(v: NexusValue): Value {
  switch (v.kind) {
    case 'Null':
      return Value.null();
    case 'Bool':
      return Value.bool(v.value);
    case 'Int':
      return Value.int(v.value);
    case 'Float':
      return Value.float(v.value);
    case 'Bytes':
      return Value.bytes(v.value);
    case 'Str':
      return Value.str(v.value);
    case 'Array':
      return Value.array(v.value.map(nexusToThunder));
    case 'Map':
      return Value.map(
        v.value.map(([k, val]) => [nexusToThunder(k), nexusToThunder(val)] as [Value, Value]),
      );
  }
}

/** Translate a Thunder `Value` back into the SDK's `NexusValue`. */
function thunderToNexus(v: Value): NexusValue {
  switch (v.kind) {
    case 'null':
      return nx.Null();
    case 'bool':
      return nx.Bool(v.value);
    case 'int':
      return nx.Int(v.value);
    case 'float':
      return nx.Float(v.value);
    case 'bytes':
      return nx.Bytes(v.value);
    case 'str':
      return nx.Str(v.value);
    case 'array':
      return nx.Array(v.value.map(thunderToNexus));
    case 'map':
      return nx.Map(
        v.value.map(([k, val]) => [thunderToNexus(k), thunderToNexus(val)] as [NexusValue, NexusValue]),
      );
  }
}

export class RpcTransport implements Transport {
  private readonly endpoint: Endpoint;
  private readonly credentials: TransportCredentials;
  private readonly connectTimeoutMs: number;
  private client: Client | null = null;
  private connectPromise: Promise<Client> | null = null;

  constructor(endpoint: Endpoint, credentials: TransportCredentials, connectTimeoutMs = 5_000) {
    this.endpoint = endpoint;
    this.credentials = credentials;
    this.connectTimeoutMs = connectTimeoutMs;
  }

  async execute(req: TransportRequest): Promise<TransportResponse> {
    const client = await this.ensureConnected();
    // Thunder's `call` returns the `Ok` payload directly and throws a typed
    // `ThunderError` (Auth/Server/Connection/Timeout/Decode/…) on `Err`; the
    // SDK's higher layers surface those to the caller unchanged.
    const result = await client.call(req.command, req.args.map(nexusToThunder));
    return { value: thunderToNexus(result) };
  }

  describe(): string {
    return `${endpointToString(this.endpoint)} (RPC)`;
  }

  isRpc(): boolean {
    return true;
  }

  async close(): Promise<void> {
    const client = this.client;
    this.client = null;
    this.connectPromise = null;
    if (client) {
      await client.close();
    }
  }

  /** Thunder config matching the Nexus RPC listener. */
  private config(): Config {
    return Config.standard()
      .withScheme('nexus')
      .withPort(this.endpoint.port)
      .withHandshake('auth_command')
      .withHelloStyle('arg_less');
  }

  /** SDK credentials → Thunder credentials (api key wins, then user/pass). */
  private thunderCredentials(): Credentials | undefined {
    if (this.credentials.apiKey) {
      return { type: 'apiKey', apiKey: this.credentials.apiKey };
    }
    if (this.credentials.username && this.credentials.password) {
      return {
        type: 'userPass',
        user: this.credentials.username,
        pass: this.credentials.password,
      };
    }
    return undefined;
  }

  /** Lazy-connect a single Thunder client, reusing it across calls. */
  private async ensureConnected(): Promise<Client> {
    if (this.client && this.client.isAlive) {
      return this.client;
    }
    if (this.connectPromise) {
      return this.connectPromise;
    }
    // Bare `host:port` endpoint sidesteps Thunder's scheme matching.
    this.connectPromise = Client.connect(
      `${this.endpoint.host}:${this.endpoint.port}`,
      this.config(),
      {
        credentials: this.thunderCredentials(),
        connectTimeoutMs: this.connectTimeoutMs,
      },
    )
      .then((client) => {
        this.client = client;
        this.connectPromise = null;
        return client;
      })
      .catch((err) => {
        this.connectPromise = null;
        throw err;
      });
    return this.connectPromise;
  }
}
